lib/Viroverse/ISLAWorksheet.pm

```diff
@@ -0,0 +1,260 @@
+
+    # Table header format
+    my $header = $xls->add_format(
+        locked      => 1,
+        bg_color    => '#FFCCCC',
+        bold        => 1,
+        color       => '#000055',
+    );
+
+    # Input data label format
+    my $label = $xls->add_format(
+        locked      => 1,
+        bold        => 1,
+        color       => '#003333',
+    );
+
+    # Locked data cell format
+    my $locked = $xls->add_format(
+        locked      => 1,
+        bold        => 1,
+        color       => '#777755',
+    );
+
+    my $locked_date = $xls->add_format();
+    $locked_date->copy($locked);
+    $locked_date->set_format_properties(
+        num_format  => 'm/d/yyyy'
+    );
+
+    # Locked data cell that carries a URL
+    my $locked_url = $xls->add_format();
+    $locked_url->copy($xls->get_default_url_format());
+    $locked_url->set_format_properties(
+        locked      => 1,
+        bold        => 0,
+        bg_color    => '#EEEEEE',
+    );
+
+    # Entry data cell format
+    my $unlocked = $xls->add_format(
+        locked      => 0,
+        bold        => 0,
+        color       => 'black',
+    );
+
+    my $unlocked_date = $xls->add_format();
+    $unlocked_date->copy($unlocked);
+    $unlocked_date->set_format_properties(
+        num_format  => 'm/d/yyyy'
+    );
+
+    # Positive PCR well
+    my $pcr_positive = $xls->add_format(
+        locked      => 0,
+        align       => 'center',
+        bg_color    => '#FFFF7F',
+        bold        => 1,
+        color       => 'black',
+    );
+    # Negative PCR well
+    my $pcr_negative = $xls->add_format(
+        locked      => 0,
+        align       => 'center',
+        bg_color    => '#CCCCCC',
+        bold        => 1,
+        color       => 'black',
+    );
+    # PCR well with invalid data
+    my $pcr_error = $xls->add_format(
+        locked      => 0,
+        bg_color    => '#990000',
+        bold        => 1,
+        color       => 'white',
+    );
+
+
+    my $sample = $self->model;
+
+    # Current write index
+    my $row = 0;
+
+    # Output the metadata table header
+    $sheet->write($row, 0, 'start_metadata', $header);
+    $sheet->write($row, 1, 'Name',           $header);
+    $sheet->write($row, 2, 'Value',          $header);
+    ++$row;
+
+    # Function to add a metadata row
+    my $add_row = sub {
+        my ($key, $name, $value, $value_format) = @_;
+
+        $sheet->write($row, 0, $key,    $label);
+        $sheet->write($row, 1, $name,   $label);
+        if (ref $value eq 'DateTime') {
+            $sheet->write_date_time(
+                $row, 2,
+                $value->strftime('%Y-%m-%dT'),
+                $value_format);
+        } else {
+            $sheet->write(
+                $row, 2,
+                $value,
+                $value_format);
+        }
+        ++$row;
+    };
+
+    # If we were given a sample URL, associate it with the sample_id row
+    $sheet->write_url($row, 2, $url) if $url;
```

Comment:
Injecting the URL from the controller is the right call but I think we should do it as an optional attribute of the `ISLAWorksheet` not a method parameter. (`has sample_url`).